### PR TITLE
WT-9500 Fix RTS to use cell time window instead of key/value timestamps

### PR DESCRIPTION
**Pull request by kommiharibabu**
RTS reads the timestamps present as part of the key/value of a history store
record to decide it's visibility. These timestamp values can be wrong when
there exists a globally visible tombstone in the update list and leads to
removal of the entire key by restoring a wrong update/tombstone combination.

To avoid RTS generating wrong update/tombstones, always read the cell time
window information to decide the HS update visibility.

### CODE REVIEW

**Comment by agorrod** [comment]
Thanks Hari - this change makes sense to me. I took a look and it appears as though all places consistently use the time window as extracted by the history store cursor after this change - which is appealing.
If that is the policy it would be nice to add in some form of checking to make sure we comply with that in the future (not suggesting as part of this work).
I added additional testing to this PR, trying to force any signal earlier if possible.

**Review by agorrod — APPROVED** [review]
Thanks @kommiharibabu - this change looks good to me.

As I mentioned somewhere, I loaded additional testing into the PR. I've analysed the results and the only non-obvious failure was WT-8967. There are other stress tests still running, but I think we could merge without waiting for everything to complete.

i.e: testing on the PR looks good to me so far.

**Comment by agorrod** [comment]
This definitely deserves at least two more reviewers before merge though...

**Review by sauclovian-wt — APPROVED** [review]
Overall LGTM - one comment that (assuming I'm not just wrong) you might want to deal with in a separate ticket.

File: src/txn/txn_rollback_to_stable.c
Comment (quchenhao) [reply src/txn/txn_rollback_to_stable.c:58]: I think it is ok to break. If U2 and U1 are already in the history store, they should be with 0 timestamp already.

**Comment by quchenhao** [comment]
@kommiharibabu Why we can't skip the hs record with globally visible tombstone in the hs cursor so we don't need to worry about this issue.

**Comment by kommiharibabu** [comment]
@quchenhao I think it is possible to ignore global visible values in the HS itself, but need to check the scenarios of recovery where RTS compares the values against the checkpoint snapshot and global visibility based on the connection level information.

**Comment by quchenhao** [comment]
@kommiharibabu In checkpoint cursor, we have implemented the way to skip globally visible tombstone from a checkpoint. We can reuse the same way.

**Comment by agorrod** [comment]
@quchenhao I'm not fully understanding your proposal - would it be hard to code it up separately, to see what it looks like?

FWIW: I quite like the approach here for now - it seems constrained to where the known bug is, but I'd be very happy to move to something encapsulated in the history store cursor moving forward. When reviewing this change, it occurred to me that it would be nice to get rid of all hs_cursor->get_key and hs_cursor->get_value calls that are outside of the history store cursor code.

OTOH: If we could see the HS cursor version it will help make the decision about which is best in the short term.

**Review by bvpvamsikrishna — APPROVED** [review]
lgtm

**Review by quchenhao — APPROVED** [review]
lgtm

**Comment by kommiharibabu** [comment]
evergreen merge

File: src/txn/txn_rollback_to_stable.c
Comment (sauclovian-wt) [reply src/txn/txn_rollback_to_stable.c:58]: I suppose it's not possible to read U2 after RTS because it's not possible for U3 to be invisible at that point. (Since you can't call RTS with a transaction active, it shouldn't be possible to read afterwards with a snapshot from before, so U3 will always be visible.)

In that case though maybe RTS should remove both U1 and U2... hmm. Not now I guess.

**Pull request merged**
